Add tests for motion module

aperezhortal opened the pull request:
Test the convergence to the actual solution of the optical flow method used.

Test that  inputs with no precipitation give a zero motion field.

Add examples to the user guide where the convergence of the optical flow methods is tested using idealized motion fields.

Fix bug with vet mask. When masked arrays where used as input, the original mask was modified unintentionally.

aperezhortal committed:
Add example to evaluate the performance of the optical flow methods

aperezhortal committed:
Fix bug with vet mask

When masked arrays where used as input, the original mask was modified unintentionally.

aperezhortal committed:
Add helper function to get a example precipitation field

aperezhortal committed:
Change title

aperezhortal committed:
Fix typos. Pep8 auto-format.

aperezhortal committed:
Add tests for motion module

aperezhortal committed:
Add example with analysis of optical flow methods convergence

aperezhortal committed:
Remove unused imports

aperezhortal committed:
Include all example files in the gallery

aperezhortal committed:
Update motion module tests

- Add zero precip tests
- Improve function documentation
- Add more parametric tests

aperezhortal committed:
Merge branch 'master' into motion_test

dnerini commented on examples/optical_flow_methods_convergence.py:
```diff
@@ -0,0 +1,363 @@
+    plt.figure(figsize=(9, 4))
+    ax = plot_precip_field(input_precip, geodata=metadata,
+                           title="Reference motion",
+                           cartopy_subplot=(1, 2, 1), map='cartopy')
```
Here `map='cartopy'` causes the doc build to fail, since cartopy is not an actual requirement, but only an optional dependency. I faced a similar problem with other examples. 
A simpler figure without any basemap will fix the issue, but we could also consider installing cartopy when building the documentation.

aperezhortal replied:
Fix in 8b6eeef

dnerini commented on examples/optical_flow_methods_convergence.py:
```diff
@@ -0,0 +1,363 @@
+importer_name = data_source["importer"]
+importer_kwargs = data_source["importer_kwargs"]
+
+# Find the input files from the archive
```
```suggestion
# Find the reference field in the archive
```

dnerini commented on examples/optical_flow_methods_convergence.py:
```diff
@@ -0,0 +1,363 @@
+fns = io.archive.find_by_date(date, root_path, path_fmt, fn_pattern, fn_ext,
+                              timestep=5, num_prev_files=0)
+
+# Read the radar composites
```
```suggestion
# Read the reference radar composite
```

dnerini commented on examples/optical_flow_methods_convergence.py:
```diff
@@ -0,0 +1,363 @@
+# This suppress nan conversion warnings in plot functions
+reference_field.data[reference_field.mask] = np.nan
+
+plot_precip_field(reference_field, geodata=metadata, title="Reference field")
```
This won't work out, since `plot_precip_field`assumes that the input field is in units of mm/h by default. Unfortunately the method doesn't support dBR, so you need to plot the field before the log transformation.

aperezhortal replied:
Fix in 8b6eeef

dnerini reviewed (changes requested):
Great contribution, @aperezhortal!
The new gallery tutorial is brilliant and sets a much higher standard for our examples ;-)

I just included few suggestions and comments, in particular there are two issues that need to be addressed before merging the PR.

dnerini committed:
Improve comments in code

Co-Authored-By: aperezhortal <andresperezcba@gmail.com>

aperezhortal commented:
Thanks @dnerini. I'm working on your suggestions. I will remove the dependency on cartopy and make plots without a base map. The map is not really important for this example.

I also found an unexpected behavior on the VET tests. The errors vary between different runs of the scripts. My guess is that it is not converging to the solution so I need have to play a little bit more with the vet options (number of iterations or minimization methods). 

**[UPDATE]**: In commit aa1c177, the VET maximum iterations was increased to give more stability to the tests.

aperezhortal committed:
Plot precip in mm/hr and remove basemap.

aperezhortal committed:
Increase VET maxiter from 100 to 150

By default, the maximum number of iteration in the VET minimization is maxiter=100.

To increase the stability of the tests to we increase this value to maxiter=150.

aperezhortal committed:
Remove ticks from axes

dnerini committed:
Merge branch 'master' into motion_test

dnerini commented on examples/optical_flow_methods_convergence.py:
```diff
@@ -0,0 +1,361 @@
+    oflow_method = motion.get_method(optflow_method_name)
+
+    computed_motion = oflow_method(precip_obs, verbose=False)
+
```
Add additional step to store precip field in mm/h so that we can plot it later.

```suggestion

precip_obs, _ = stp.utils.dB_transform(precip_obs, inverse=True)

```

dnerini commented on examples/optical_flow_methods_convergence.py:
```diff
@@ -0,0 +1,361 @@
+
+    computed_motion = oflow_method(precip_obs, verbose=False)
+
+    precip_data, _ = stp.utils.dB_transform(precip_obs.max(axis=0),
```
```suggestion
    precip_data = precip_obs.max(axis=0)
```

dnerini commented on examples/optical_flow_methods_convergence.py:
```diff
@@ -0,0 +1,361 @@
+    quiver(ideal_motion, step=25, ax=ax)
+
+    plt.subplot(1, 2, 2)
+    ax = plot_precip_field(input_precip, title="Retrieved motion")
```
```suggestion
    ax = plot_precip_field(precip_obs[0], title="Retrieved motion")
```

dnerini commented on examples/optical_flow_methods_convergence.py:
```diff
@@ -0,0 +1,361 @@
+    # Compare retrieved motion field with the ideal one
+    plt.figure(figsize=(9, 4))
+    plt.subplot(1, 2, 1)
+    ax = plot_precip_field(input_precip, title="Reference motion")
```
```suggestion
    ax = plot_precip_field(precip_obs[0], title="Reference motion")
```

dnerini commented on examples/optical_flow_methods_convergence.py:
```diff
@@ -0,0 +1,361 @@
+    computed_motion = oflow_method(precip_obs, verbose=False)
+
+    precip_data, _ = stp.utils.dB_transform(precip_obs.max(axis=0),
+                                            inverse=True)
```
```suggestion
```

dnerini reviewed (changes requested):
Looks great to me, I have just included a last suggestion in order to correctly plot the reference field in the results (it is the same problem you had with the first figure before).

dnerini committed:
Fix precip plot units

Co-Authored-By: aperezhortal <andresperezcba@gmail.com>

aperezhortal committed:
Fix indentation

dnerini commented:
Brilliant!
